### PR TITLE
refs codership/galera#367 use system ASIO if available

### PR DESCRIPTION
**Pull request by temeo**
Made debian build to depend on libasio-dev and enhanced SConstruct
to check if system provides ASIO devel library and if it is recent
enough (> 1.10.1).

If ASIO devel library is not installed or is not recent enough,
build falls back to use of embedded copy.

Also reduced the scope of compiler warning suppressions when
system ASIO is used and fixed compiler warnings.

### CODE REVIEW

File: debian/control [hunk 1]
```diff
@@ -9,7 +9,8 @@ Build-Depends: check,
                libboost-dev (>= 1.41),
                libboost-program-options-dev (>= 1.41),
                libssl-dev,
-               scons (>= 2)
+               scons (>= 2),
+               libasio-dev
```
Comment (ayurchen) [inline debian/control:13]: Putting libasio-dev above scons would have made a patch a tad smaller ;) Also, if this is made a build dependency on debian, whoud not it make sense to specify a minimum version as well? I know, we check for it during build anyway, but it is further down the line.

File: galera/src/ist_proto.hpp [hunk 2]
```diff
@@ -62,6 +74,8 @@ namespace galera
                 len_    (len    )
             { }
 
+            ~Message() { }
```
Comment (ayurchen) [inline galera/src/ist_proto.hpp:77]: why was this needed? It used to build before.

File: galerautils/src/gu_asio.cpp [hunk 3]
```diff
@@ -5,6 +5,7 @@
 #include "gu_config.hpp"
 #include "gu_asio.hpp"
 
+#include <boost/bind.hpp>
```
Comment (ayurchen) [inline galerautils/src/gu_asio.cpp:8]: Should not it be wrapped in
    #fidef HAVE_SYSTEM_ASIO
    #endif
?

**Comment by ayurchen** [comment]
On a philosophical note - libasio-dev seems to depend on boost. So I'd call the macro HAVE_BOOST_ASIO rather than HAVE_SYSTEM_ASIO - more descriptive.

**Comment by ayurchen** [comment]
Overall only minor notes except for boost bind.h header which is not ifdefed

File: debian/control (hunk 1 above)
Comment (temeo) [inline debian/control:13]: Minimum version requirement would make a build fail on older debian/ubuntu targets where recent enough libasio-dev is not available. This looked like good enough compromise: Have a build dependency on libasio-dev library, but don't use it if it is not recent enough.

File: galera/src/ist_proto.hpp (hunk 2 above)
Comment (temeo) [inline galera/src/ist_proto.hpp:77]: I tried to make code build with -Weffc++ which required virtual destructor. However, this broke protocol due to dependency on Message class size, so I removed the virtual but left explicit destructor.

File: galerautils/src/gu_asio.cpp (hunk 3 above)
Comment (temeo) [inline galerautils/src/gu_asio.cpp:8]: Code uses `boost::bind` explicitly, so I think it is good to include `boost/bind.hpp` explicitly too and not to rely on other headers on this.

**Comment by temeo** [comment]
I would still keep HAVE_SYSTEM_ASIO, HAVE_BOOST_ASIO should be reserved to the case if we will use asio provided by boost libraries at some point.

**Comment by ayurchen** [comment]
> I would still keep HAVE_SYSTEM_ASIO, HAVE_BOOST_ASIO should be reserved to the case if we will use asio provided by boost libraries at some point.

Isn't it the asio that "is provided" by boost already? Otherwise why would it depend on boost? - that's what I thought.
Anyway, good to merge.

**Comment by temeo** [comment]
> Isn't it the asio that "is provided" by boost already? Otherwise why would it depend on boost? - that's what I thought.

Boost provides asio, but there is also standalone libasio which we use for historical reasons. Asio provided by boost lives in different namespace than standalone, so using it would require a bit more work, maybe worth of separate patch.

Edit: Using asio provided by boost also changes galera library linkage by requiring dynamic linking against libboost_system.
